File: app/repositories/position_repository.py
from sqlalchemy import select

from app.models.position import Position


class PositionRepository:

    def __init__(
        self,
        db
    ):
        self.db = db

    async def create(
        self,
        position: Position
    ):

        self.db.add(position)

        await self.db.commit()

        await self.db.refresh(position)

        return position

    async def get_by_symbol(
        self,
        symbol: str
    ):

        result = await self.db.execute(
            select(Position).where(
                Position.symbol == symbol
            )
        )

        return result.scalar_one_or_none()

    async def get_all(self):

        result = await self.db.execute(
            select(Position)
        )

        return result.scalars().all()

    async def delete(
        self,
        position: Position
    ):

        await self.db.delete(position)

        await self.db.commit()

    async def update(self):

        await self.db.commit()

File: backend/src/repositories/positionRepository.ts
import { prisma } from "../core/database";
import { Position } from "@prisma/client";

export class PositionRepository {
  static async create(data: {
    id: string;
    symbol: string;
    quantity: number;
    average_price: number;
    stop_loss?: number | null;
    take_profit?: number | null;
    unrealized_pnl?: number;
  }): Promise<Position> {
    return prisma.position.create({
      data,
    });
  }

  static async getBySymbol(symbol: string): Promise<Position | null> {
    return prisma.position.findUnique({
      where: {
        symbol,
      },
    });
  }

  static async getAll(): Promise<Position[]> {
    return prisma.position.findMany();
  }

  static async delete(symbol: string): Promise<void> {
    await prisma.position.delete({
      where: {
        symbol,
      },
    });
  }

  static async update(symbol: string, data: {
    quantity?: number;
    average_price?: number;
    stop_loss?: number | null;
    take_profit?: number | null;
    unrealized_pnl?: number;
  }): Promise<Position> {
    return prisma.position.update({
      where: {
        symbol,
      },
      data,
    });
  }
}
export default PositionRepository;
